lgc/patch/PatchBufferOp.cpp

```diff
@@ -64,129 +79,278 @@ PreservedAnalyses PatchBufferOp::run(Function &function, FunctionAnalysisManager
+    visitor.visit(*this, *inst);
+  assert(m_postVisitInsts.empty());
+}
+
+// =====================================================================================================================
+// Get the descriptor info describing whether the descriptor is invariant and/or divergent.
+//
+// This function resolves invariance and divergence to a "known" state if necessary.
+//
+// @param desc : the descriptor
+BufferOpLowering::DescriptorInfo BufferOpLowering::getDescriptorInfo(Value *desc) {
+  DescriptorInfo di = m_descriptors[desc];
+  if (di.invariant.has_value() && di.divergent.has_value())
+    return di;
+
+  // Resolve by first finding all necessary roots and then performing an aggressive data flow fixed point iteration,
+  // i.e. start with the tacit assumption that all descriptors are uniform and invariant.
+  DenseSet<Value *> seen;
+  SmallVector<Value *> searchWorklist;
+  SmallVector<Value *> propagationWorklist;
+  Value *current = desc;
+  for (;;) {
+    if (seen.insert(current).second) {
+      auto &di = m_descriptors[current];
+
+      if (!di.invariant.has_value() || !di.divergent.has_value()) {
+        if (auto *phi = dyn_cast<PHINode>(current)) {
+          for (Value *incoming : phi->incoming_values())
+            searchWorklist.push_back(incoming);
+        } else if (auto *select = dyn_cast<SelectInst>(current)) {
+          searchWorklist.push_back(select->getTrueValue());
+          searchWorklist.push_back(select->getFalseValue());
+        } else {
+          // Make conservative assumptions for unhandled instructions.
+          bool isConstant = isa<Constant>(current);
+          auto &di = m_descriptors[current];
+          if (!di.invariant.has_value())
+            di.invariant = isConstant;
+          if (!di.divergent.has_value()) {
+            // TODO: This would be entirely unnecessary if we had updatable divergence info.
+            di.divergent = !isConstant;
+          }
+        }
+      }
+
+      if (!di.invariant.value_or(true) || di.divergent.value_or(false))
+        propagationWorklist.push_back(current);
+    }
+
+    if (searchWorklist.empty())
+      break;
+
+    current = searchWorklist.pop_back_val();
   }
-  m_postVisitInsts.clear();
 
-  const bool changed = (!m_replacementMap.empty());
+  // Fixed-point iteration to propagate "variant" and "divergent" flags.
+  while (!propagationWorklist.empty()) {
+    current = propagationWorklist.pop_back_val();
+    auto diIt = m_descriptors.find(current);
+    assert(diIt != m_descriptors.end());
+    DescriptorInfo di = diIt->second;
+
+    for (User *user : current->users()) {
+      // Make a reasonable effort not to "leak" into instructions we don't understand (e.g., if a pointer / descriptor
+      // ended up in an aggregate). Some of these cases could perhaps be handled in a conservative way, but it seems
+      // unlikely to be necessary in practice.
+      if (!isa<PHINode>(user) && !isa<SelectInst>(user) && !seen.count(user))
```

Comment:
Maybe I missed something, but I don't see how a select instruction was properly handled. A select with divergent condition will get divergent descriptor while a select with uniform condition would get uniform descriptor if the selected descriptors are uniform. Could you help explain?

Reply:
There was indeed something missing here. I've added a check for select instructions with divergent condition, to make sure that they are marked as divergent.